File: ts/src/merge_sort.ts
#!/usr/bin/env ts-node

/**
 * Merge sort implementation in TypeScript
 * Time complexity: O(n log n)
 * Space complexity: O(n)
 *
 * @param arr The array to sort
 * @returns The sorted array
 */
export function mergeSort(arr: number[]): number[] {
    if (arr.length <= 1) return arr;

    // Split array into two halves
    const mid = Math.floor(arr.length / 2);
    const left = arr.slice(0, mid);
    const right = arr.slice(mid);

    // Recursively sort both halves
    const sortedLeft = mergeSort(left);
    const sortedRight = mergeSort(right);

    // Merge the sorted halves
    return merge(sortedLeft, sortedRight);
}

/**
 * Helper function to merge two sorted arrays
 *
 * @param left First sorted array
 * @param right Second sorted array
 * @returns Merged sorted array
 */
function merge(left: number[], right: number[]): number[] {
    const result: number[] = [];
    let i = 0;
    let j = 0;

    // Compare elements from both arrays and merge them
    while (i < left.length && j < right.length) {
        if (left[i] <= right[j]) {
            result.push(left[i]);
            i++;
        } else {
            result.push(right[j]);
            j++;
        }
    }

    // Copy remaining elements from left array
    result.push(...left.slice(i));

    // Copy remaining elements from right array
    result.push(...right.slice(j));

    return result;
}

// Check if file path is provided
if (process.argv.length < 3) {
    console.error("Please provide a file path as an argument");
    process.exit(1);
}

// Read numbers from file
import * as fs from 'fs';
const numbers = fs.readFileSync(process.argv[2], 'utf8')
    .split('\n')
    .filter(line => line.trim())
    .map(Number);

// Sort the array
const sortedNumbers = mergeSort(numbers);

// Write sorted numbers back to file
fs.writeFileSync('sorted_data.txt', sortedNumbers.join('\n'));

File: csharp/src/MergeSort.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

namespace MergeSort
{
    /// <summary>
    /// Implementation of merge sort algorithm in C#
    /// Time complexity: O(n log n)
    /// Space complexity: O(n)
    /// </summary>
    public class MergeSort<T> where T : IComparable<T>
    {
        /// <summary>
        /// Sorts an array using merge sort algorithm
        /// </summary>
        /// <param name="arr">The array to sort</param>
        /// <returns>The sorted array</returns>
        public static T[] Sort(T[] arr)
        {
            if (arr == null || arr.Length <= 1)
                return arr;

            // Split array into two halves
            int mid = arr.Length / 2;
            T[] left = arr.Take(mid).ToArray();
            T[] right = arr.Skip(mid).ToArray();

            // Recursively sort both halves
            return Merge(Sort(left), Sort(right));
        }

        /// <summary>
        /// Merges two sorted arrays into a single sorted array
        /// </summary>
        /// <param name="left">First sorted array</param>
        /// <param name="right">Second sorted array</param>
        /// <returns>Merged sorted array</returns>
        private static T[] Merge(T[] left, T[] right)
        {
            var result = new List<T>(left.Length + right.Length);
            int i = 0, j = 0;

            // Compare elements from both arrays and merge them
            while (i < left.Length && j < right.Length)
            {
                if (left[i].CompareTo(right[j]) <= 0)
                {
                    result.Add(left[i++]);
                }
                else
                {
                    result.Add(right[j++]);
                }
            }

            // Copy remaining elements from left array
            while (i < left.Length)
            {
                result.Add(left[i++]);
            }

            // Copy remaining elements from right array
            while (j < right.Length)
            {
                result.Add(right[j++]);
            }

            return result.ToArray();
        }
    }

    class Program
    {
        static void Main(string[] args)
        {
            if (args.Length < 1)
            {
                Console.Error.WriteLine("Please provide a file path as an argument");
                Environment.Exit(1);
            }

            try
            {
                // Read numbers from file
                var numbers = File.ReadAllLines(args[0])
                    .Where(line => !string.IsNullOrWhiteSpace(line))
                    .Select(line => int.Parse(line.Trim()))
                    .ToArray();

                // Sort the array
                var sortedNumbers = MergeSort<int>.Sort(numbers);

                // Write sorted numbers back to file
                File.WriteAllLines("sorted_data.txt",
                    sortedNumbers.Select(n => n.ToString()));
            }
            catch (Exception ex)
            {
                Console.Error.WriteLine($"Error: {ex.Message}");
                Environment.Exit(1);
            }
        }
    }
}
